### ISSUE TITLE
FileLogHandler - Implement Rotating Log Handling and Culling

### ISSUE THREAD

**Issue by mbaker3** [decline-cookies/anvil-csharp-core#108]
When `FileLogHandler` is being used with the `append` parameter set to true the individual log file will grow indefinitely.

This does not necessarily need to be implemented in FileLogHandler. Whoever tackles this task can decide the best course.

Implement a mechanism to:
 - Move to a new log file when it reaches a specific size
 - Cull old log files

This should work similarly to [logrotate](https://www.tecmint.com/install-logrotate-to-manage-log-rotation-in-linux/) on linux.
 - The initial file name is as defined by the developer. `myLog.log`
 - When the initial log file fills up it should be renamed to `myLog.1.log` and a new `myLog.log` created.
 - When the log file fills up again all logs should be rotated by 1 and the current log file becoming the new `myLog.1.log.

This results in a log collection with a maximum rotation count of 3 looking like:
 - `myLog.log` - The log file currently being written to
 - `myLog.1.log` - The next newest log file where the last entry is the log emitted immediately before the first entry in `myLog.log`. When `myLog.log` fills up this will be renamed to `myLog.2.log`
 - `myLog.2.log`
 - `myLog.3.log` - The oldest log file. When `myLog.log` fills this file will be culled.

The consuming developer should have parameters exposed (with sane defaults) to configure:
 - Threshold size to roll over to a new log
 - Maximum number of times a log may be rotated.

**Comment by tjvezina** [comment]
I don't see the harm in making these options available to non-appended logs too - maybe you've got a lot of logs going on and you'd like them rotated, even if they're wiped out and restarted each run.

That does complicate the implementation a bit though - would it be worth setting that up for the increased flexibility, or should it only apply to appended logs? In the latter case, I'm thinking a `RotatedFileLogHandler` derived from `FileLogHandler`.

**Comment by tjvezina** [comment]
Also, in a completely biased move based on a single recent experience (Plex logs), I'm thinking the default limits could be 10MB / 5 files. Thoughts?

**Comment by jkeon** [comment]
I would maybe take it further and say it's good policy to keep a history (however short) of the logs regardless. 

Whether it's append or not, in a situation where you need to look at the log, it's probably helpful to be able to look at the previous session's log as well for comparison. That might make the implementation simpler.

With the default limits, I don't have a strong opinion, I'm fine with the 10MB / 5 files default. I would imagine the API will make it trivial for each application to set those limits as it best suits them.

**Comment by tjvezina** [comment]
> Whether it's append or not, [...] it's probably helpful to be able to look at the previous session's log

How would that work exactly? If you create a `new FileLogHandler("myLog.txt", append: false);` the StreamWriter will clear "myLog.txt" before writing to it. Are you suggesting it should save the previous log each time instead, before starting a new one?

**Comment by jkeon** [comment]
Hmmn I'm not 100% sure how it works today but yeah I was under the impression that you set what your "name" of the log is. "myLog.txt" for example.

If there's no history, then it just creates "myLog.txt". 

If the file already exists, it saves the existing one as "myLog.1.txt" and then creates the new one. And so on. 

I don't want to derail this issue or discussion but thinking out loud about the usage of logs for debugging, i'm wondering if the filenames should always have a date/time stamp in them as well?  I guess with append that complicates things since it could span more than one session. What's the reason for appending across multiple sessions actually?

**Comment by mbaker3** [comment]
> 10MB / 5 files

Good defaults.

### Re: Rotating Non-appended logs:
I like the idea of rotating logs for a single session if it gets too long. This has the knock on affect of allowing the consumer to constrain the amount of space log files take up which may be a requirement in their run context.

I don't think this feature needs to be implemented as part of this task but will leave it to the implementing developer to decide whether it's easy to include in scope or not.

I don't think non-append mode logs should be rotated between sessions. In my mind, the use case for non-append logs is for things like "previous session" log files that always represent the last run.
Also, If you rotate on each new session then you end up with a confusing result where you don't know if a rotated file is a different run session or a long session that over-ran the single file limit.

If you as the developer want to keep historical logs but for some reason don't want to use append mode (why?) then  you can tag your log file name with a session unique identifier (Ex: `myLogs-{DateTime.Now}.log`).

Maybe a nice to have is that the Log rotation logic can be used on its own without a specific log handler?
Not sure how exactly that would work but something along the lines of providing a file name prefix, path, rotate parameters and then it will mutate the directory for you?

**Comment by jkeon** [comment]
> I like the idea of rotating logs for a single session if it gets too long. This has the knock on affect of allowing the consumer to constrain the amount of space log files take up which may be a requirement in their run context.

I don't have any problems with this, but I think it will be confusing if there's no session identifier. 

- myLog.txt
- myLog.1.txt
- myLog.2.txt
- myLog.3.txt
- myLog.4.txt

Those logs contain three sessions in append mode and had exceeded the size threshold. You'll have to just guess and open them and do a CTRL-F for some log you know is at the beginning of the session? Not ideal. 

I was suggesting something like this (actual format TBD)

- myLog-2022-09-07-11-34-12.txt
- myLog-2022-09-07-11-34-12.1.txt
- myLog-2022-09-07-11-04-45.txt
- myLog-2022-09-07-09-12-26.txt
- myLog-2022-09-07-09-12-26.1.txt

Much easier to see that there were three sessions. One at 9:12, one at 11:04 and one at 11:34 with both 9:12 and 11:34 exceeding the threshold and being rotated. I can easily open to the start of my log for the session I want to debug.

> I don't think non-append mode logs should be rotated between sessions. In my mind, the use case for non-append logs is for things like "previous session" log files that always represent the last run.

I'm not sure I fully understand this. How would this be used? Wouldn't your code always just be logging the current session?

> Also, If you rotate on each new session then you end up with a confusing result where you don't know if a rotated file is a different run session or a long session that over-ran the single file limit.

Agreed but I think it's confusing right now anyway with no way to denote between sessions. See my above comment on the datetimestamp.

> If you as the developer want to keep historical logs but for some reason don't want to use append mode (why?) then you can tag your log file name with a session unique identifier (Ex: myLogs-{DateTime.Now}.log).

This comment is making me feel like I'm looking at this one level higher than you. In which case maybe all of my comments are moot. 

Are we saying that for each "logfile" it would be split across 5 parts with each part being up to 10mb and that the name of the logfile is the "session"?

Assume I run the app once for 30 days in a row. I would have 150 files, each at 10mb assuming they all "filled up" completely. Its up to me as the app developer to cull those as I see fit? 

In which case yeah, ignore everything I said, I was thinking that this would encompass more functionality and we were saying you could only have the 5 logfiles total.

**Comment by mbaker3** [comment]
> Those logs contain three sessions in append mode and had exceeded the size threshold. You'll have to just guess and open them and do a CTRL-F for some log you know is at the beginning of the session? Not ideal.
>
> I was suggesting something like this (actual format TBD)
> ...

I'm hesitant to move in this direction. Do we think we know enough about log file consumption to go against the grain of unix style log rotation?

Personally, I'd prefer all of the logging to be continuous in as few files as possible. Log files can be easily split on a predefined session start identifier when reading if desired but more often than not the context of the previous session is important to debugging the current session. I'd prefer to be able to just review the most recent log file and get an idea of what has been going over the recent runtime. Whether that's part of a single session or multiple short sessions.

Reading further it's clear this is an application level choice to be made and probably not something we should be baking into the anvil log rotation logic.



>> I don't think non-append mode logs should be rotated between sessions. In my mind, the use case for non-append logs is for things like "previous session" log files that always represent the last run.
>
> I'm not sure I fully understand this. How would this be used? Wouldn't your code always just be logging the current session?

Yea, a lot of log debugging during development that you're going to do is on the previous/current session and you don't care about the previous sessions. If you always log to one "last session" file without appending then you can keep that file open in a text editor and easily see the current logs cleared of any history. Just as you would in an IDE.

Maybe the gap here is that I envision an application with robust logging that makes use of multiple File log handlers each with their own specific purpose.

> Are we saying that for each "logfile" it would be split across 5 parts with each part being up to 10mb and that the name of the logfile is the "session"?

I'm saying it's up to the developer to decide what they want for their given application. If you want to write a per session log file, go for it, but if you want to keep using the same files, go for it. We've got your back.

> Assume I run the app once for 30 days in a row. I would have 150 files, each at 10mb assuming they all "filled up" completely. Its up to me as the app developer to cull those as I see fit?

If you have a single app session that lasts 4o days and generates 150 * 10mb of log data then only the most recent 5 10mb log files would exist.

**Comment by jkeon** [comment]
Yeah ok, sounds good. Looks like my concerns are handled at the app level.

**Comment by tjvezina** [comment]
I think we're conflating two different purposes for having multiple files - on the one hand, there's file "rotation" which is meant to break a single file into multiple parts to prevent any one from growing to large, and on the other hand there's keeping a history of previous logs for later reference.

I think the `FileLogHandler` should take care of file size limiting/rotation for you (which will be configurable), and keeping a history is up to you, which can be achieved as has been demonstrating by appending a timestamp of some kind to your file name when the log handler is created.

So, focusing on the file rotation part only, what are the use cases?

### 1) New log each session, replace old (`append: false, rotate: false`)
I just want a dead-simple logger that dumps all my logs to a file each run, replacing the previous one.

### 2) New log each session, rotate files (`append: false, rotate: true`)
Each session, the previous log files are deleted (including rotations) and we start fresh. As the file size limit is reached, logs are rotated, and if a file count limit is set the oldest is deleted.

### 3) Continue adding to the existing log file, no rotation (`append: true, rotate: false`)
Plain and simple, one mega log file.

### 4) Continue adding to the existing log file, rotate files (`append: true, rotate: true`)
Log files are a continuous history - no clear start/end of sessions, but you chose this with `append: true`.

---

Any of these can be combined with a unique timestamp in the log file name to create a history. The question is, should I bother with Use Case 2 or only support file rotation for appended logs?

Also, maybe I'm wrong and the FileLogHandler should have built-in functionality for appending a unique timestamp to your log file name?

**Comment by jkeon** [comment]
Yes that's my fault, I was thinking of it from the history side which should be handled a layer above in the app.

> The question is, should I bother with Use Case 2 or only support file rotation for appended logs?

While I think @mbaker3 is saying that Use Case 2 isn't something that he expects to come up regularly, if the API is those two booleans, it feels weird to have one combination that doesn't do anything? 

> Also, maybe I'm wrong and the FileLogHandler should have built-in functionality for appending a unique timestamp to your log file name?

No you were right, that can be built through the app layer.

**Comment by tjvezina** [comment]
If we don't want to support Use Case 2, I would leave FileLogHandler as-is (with support for cases 1 and 3) and add a `RotatedFileLogHandler : FileLogHandler` (which always uses append) to handle case 4.

Otherwise, I'll modify `FileLogHandler` to support all cases, with whatever API fits best.

**Comment by mbaker3** [comment]
Haven't looked at the PR yet but this all sounds good. I was hoping we could keep the rotation logic separate from the `FileLogHandler` but it may be more trouble than it's worth.

**Comment by tjvezina** [comment]
I can separate it out, but you can still say `new FileLogHandler(shouldRotate: false)` - if you prefer a separate `RotatedFileLogHandler` (or however you want to name it, I can do that too

**Comment by mbaker3** [comment]
Left some comments on the PR. What I'm imagining may be a larger effort than the scope of the current work. I'm completely fine keeping it narrow for now and expanding later. Can continue discussing on PR :)
